sdk/java-sdk-spring/src/main/scala/kalix/spring/impl/KalixSpringApplication.scala

```diff
@@ -368,88 +299,106 @@ case class KalixSpringApplication(applicationContext: ApplicationContext, config
 
   def port: Int = kalixRunner.configuration.userFunctionPort
 
-  /* Each component may have a creation context passed to its constructor.
-   * This method checks if there is a constructor in `clz` that receives a `context`.
-   */
-  private def hasContextConstructor(clz: Class[_], contextType: Class[_]): Boolean =
-    clz.getConstructors.exists { ctor =>
-      ctor.getParameterTypes.contains(contextType)
-    }
-
-  private def actionProvider[A <: Action](clz: Class[A]): ActionProvider[A] =
-    ReflectiveActionProvider.of(
-      clz,
-      messageCodec,
-      context => {
-        if (hasContextConstructor(clz, classOf[ActionCreationContext]))
-          ActionCreationContextFactoryBean.set(context)
-
-        val webClientProviderHolder = WebClientProviderHolder(context.materializer().system)
-
-        setKalixClient(clz, webClientProviderHolder)
-        setComponentClient(clz, webClientProviderHolder)
+  private def webClientProvider(context: Context) = {
+    val webClientProviderHolder = WebClientProviderHolder(context.materializer().system)
+    webClientProviderHolder.webClientProvider
+  }
 
-        if (hasContextConstructor(clz, classOf[WebClientProvider])) {
-          val webClientProvider = webClientProviderHolder.webClientProvider
-          WebClientProviderFactoryBean.set(webClientProvider)
-        }
+  private def wiredInstance[T](clz: Class[T])(partial: PartialFunction[Class[_], Any]): T = {
+    // only one constructor allowed
+    // TODO: validate this
+    wiredInstance(clz.getDeclaredConstructors.head.asInstanceOf[Constructor[T]])(partial)
+  }
 
-        kalixBeanFactory.getBean(clz)
-      })
+  private def wiredInstance[T](constructor: Constructor[T])(partial: PartialFunction[Class[_], Any]): T = {
+
+    // Note that this function is total because it will always return a value (even if null)
+    // last case is a catch all that lookups in the applicationContext
+    val totalWireFunction: PartialFunction[Class[_], Any] =
+      partial.orElse {
+        // block wiring of clients into anything that is not an Action or Workflow
+        // NOTE: if they are allowed, 'partial' should already have a matching case for them
+        case p if p == classOf[KalixClient] =>
+          new BeanCreationException(
+            s"KalixClient cannot be inject into a [${constructor.getDeclaringClass.getSimpleName}]")
```

Comment:
It should just create and return the exception, not throw it?

Reply:
Oh god! I don't know how to code anymore.